ffmpeg/ffmpeg_test.go

```diff
@@ -1512,3 +1512,60 @@ func TestTranscoder_FormatOptions(t *testing.T) {
 		t.Error("Did not get expected error with invalid format ", err)
 	}
 }
+
+// test short segments
+func shortSegments(t *testing.T, accel Acceleration, fc int) {
+	run, dir := setupTest(t)
+	defer os.RemoveAll(dir)
+
+	cmd := `
+    # generate segments with fc-frames
+    cp "$1/../transcoder/test.ts" .
+	frame_count=%d
+
+    ffmpeg -loglevel warning -ss 0 -i test.ts -c copy -frames:v $frame_count -copyts short0.ts
+    ffmpeg -loglevel warning -ss 2 -i test.ts -c copy -frames:v $frame_count -copyts short1.ts
+    ffmpeg -loglevel warning -ss 4 -i test.ts -c copy -frames:v $frame_count -copyts short2.ts
+    ffmpeg -loglevel warning -ss 6 -i test.ts -c copy -frames:v $frame_count -copyts short3.ts
+
+    ffprobe -loglevel warning -count_frames -show_streams -select_streams v short0.ts | grep nb_read_frames=$frame_count
+    ffprobe -loglevel warning -count_frames -show_streams -select_streams v short1.ts | grep nb_read_frames=$frame_count
+    ffprobe -loglevel warning -count_frames -show_streams -select_streams v short2.ts | grep nb_read_frames=$frame_count
+    ffprobe -loglevel warning -count_frames -show_streams -select_streams v short3.ts | grep nb_read_frames=$frame_count
+  `
+	run(fmt.Sprintf(cmd, fc))
+
+	tc := NewTranscoder()
+	defer tc.StopTranscoder()
+	for i := 0; i < 4; i++ {
+		fname := fmt.Sprintf("%s/short%d.ts", dir, i)
+		t.Log("fname ", fname)
+		in := &TranscodeOptionsIn{Fname: fname, Accel: accel}
+		out := []TranscodeOptions{{Oname: dir + "/out.ts", Profile: P144p30fps16x9, Accel: accel}}
+		res, err := tc.Transcode(in, out)
+		if err != nil {
+			t.Error(err)
+		}
+		if fc != res.Decoded.Frames {
+			t.Error("Did not decode expected number of frames: ", res.Decoded.Frames)
+		}
+		if 0 == res.Encoded[0].Frames {
+			// not sure what should be a reasonable number here
+			t.Error("Did not encode any frames: ", res.Encoded[0].Frames)
+		}
+	}
+
+	// Also test:
```

Comment:
Would be good to have these tests here as well.

Reply:
I agree. This PR is already covering a lot of things though, will it be OK if I cut a separate issue for this?

Reply:
Would rather incorporate those tests now, because if the tests surface issues then we can address them prior to merging this PR. Especially because the changes here are relatively fraught, so I'd prefer to be certain that the behavior is as expected. Also, from experience, punting on tests usually means they'll never get done until we're neck deep in a bug that the test would have caught.

Reply:
I have updated the commit with tests for all the cases.

>  //  stream copy (both in conjunction with transcoding and standalone)

I did not exactly understand what the difference here was. I've implemented the stream copy and drop tests within the ShortSegments one after closing/re-opening the transcoder.

Reply:
Thank you - the new tests look great!

> I did not exactly understand what the difference here was.

The code paths are a bit different if the user is doing stream copy and transcoding in different renditions.

For the streamcopy-only (or streamcopy+drop), the decoder and encoder won't be initialized at all. The packet is demuxed then sent straight to the muxer.

If doing streamcopy for one output, and transcoding another output, then the normal flow is basically in place, but the copied stream will also skip the encoder.

I thought it might be good to sanity check there weren't any weird interactions when mixing those types of outputs. We do have `TestTranscoder_StreamCopy` but that's more of a regular, single-segment test case with a normal length and fps video.

Something like this wild config could exercise a number different things at once. What do you think?

```golang
profile_low_fps = P144p30fps16x9
p.Framerate = fc // use the input frame count as the output fps, why not
profile_passthrough_fps = P144p30fps16x9
p.Framerate = 0
[]TranscodeOptions{{
  Oname: dir+"/low_fps.ts"
  Profile: profile_low_fps,
  AudioEncoder: ComponentOptions{Name:"copy"},
},{
  Oname: dir+"/copy_all.ts",
  VideoEncoder: ComponentOptions{Name:"copy"},
  AudioEncoder: ComponentOptions{Name:"drop"},
},{
  Oname: dir+"/passthrough.ts",
  Profile: profile_passthrough_fps,
}}
```

(We should also keep the standalone tests that have already been written here, because those are very good to have.)

Reply:
this is very wacky, i love it, added it.